### PR TITLE
Feat/stakes api

### PR DESCRIPTION
**Pull request by mwamedacen**
# Description
This exposes 2 new endpoints
- `GET /stakes` -> all stakes for all stakers for all staking contracts (sPSPs and stkPSPBpt)
- `GET /stakes/:address` -> all stakes for one staker for all staking contracts (sPSPs and stkPSPBpt)

# How to test
start with `yarn dev` 

http://localhost:3232/stakes
http://localhost:3232/stakes/0x81ea04c179b82d31662994a68b2582ad9d860f88

or 

- https://api.staging.paraswap.io/staking/stakes
- https://api.staging.paraswap.io/staking/stakes/0x81ea04c179b82d31662994a68b2582ad9d860f88

closes: FRNT-147

### CODE REVIEW

File: src/router.ts [hunk 1]
```diff
@@ -142,6 +147,21 @@ export default class Router {
...
+        const account = req.params.account;
+        const totalPSPStakedInAllStakingPrograms =
+          await StakingService.getInstance().getPSPStakesAllPrograms(account);
+
+        return res.json(totalPSPStakedInAllStakingPrograms);
+      } catch (e) {
+        logger.error(req.path, e)
+        return res
+          .status(403)
+          .send({ error: 'stakes could not been retrieved for user' });
```
Comment (samthomson) [inline src/router.ts:161]:
```suggestion
          .send({ error: 'stakes could not be retrieved for user' });
```

File: src/lib/staking.ts [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    if (!this.instance) {
+      this.instance = new StakingService();
+    }
+
+    return this.instance;
+  }
+
+  chainId = CHAIN_ID_MAINNET; // all staking programs are only available on ethereum mainnet
+  multicallContract: Contract;
+  sPSPIFace: Interface;
```
Comment (samthomson) [inline src/lib/staking.ts:45]:
```suggestion
  sPSPIface: Interface;
```
(optional)
perhaps to be consistent.

File: src/lib/staking.ts [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+  totalPSPStaked: T;
+  descr: {
+    totalPSPStakedInSPSP: T;
+    totalPSPStakedInSafetyModule: T;
+  };
+};
+
+export const SPSPAddresses = PoolConfigsMap[CHAIN_ID_MAINNET].filter(
+  p => p.isActive,
+).map(p => p.address.toLowerCase());
```
Comment (samthomson) [inline src/lib/staking.ts:30]: we make this filtering elsewhere I think, perhaps an `activeSPSPAddresses` const in `./pool-info` could be usedful.

File: src/lib/staking.ts [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    }
+
+    return this.instance;
+  }
+
+  chainId = CHAIN_ID_MAINNET; // all staking programs are only available on ethereum mainnet
+  multicallContract: Contract;
+  sPSPIFace: Interface;
+  bVaultIface: Interface;
+  erc20Iface: Interface;
```
Comment (samthomson) [inline src/lib/staking.ts:47]: I believe we're conventionally using ERC not erc, but feel free to ignore me :sweat_smile:

**Review by samthomson — APPROVED** [review]

File: src/lib/staking.ts (hunk 2 above)
Comment (mwamedacen) [reply src/lib/staking.ts]: outdated

File: src/lib/staking/safety-module-helper.ts [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+          [SAFETY_MODULE_ADDRESS]: totalPSPStaked.toString(),
+        }),
+      },
+    };
+  };
+
+  /// functions to fetch multiple stakers data efficiently
+  async fetchStkPSPBptStakers(
+    blockNumber?: number,
+  ): Promise<DataByAccount<bigint>> {
```
Comment (mwamedacen) [inline src/lib/staking/safety-module-helper.ts:91]: I must admit I've been a little juggling between bigint and bigNumber.
In a near future I'd like to move the whole code GRP dependant code to bigint but I'm a little bit afraid of the precision loss so I prefer not to for now and keep 🤹

**Review by samthomson — APPROVED** [review]

File: src/lib/staking/safety-module-helper.ts [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+      stkPSPBalanceBN,
+    ] = await Promise.all([
+      this.fetchStkPSPBPtState(),
+      this.safetyModuleAsERC20.balanceOf(account) as Promise<EthersBN>,
+    ]);
+
+    const stkPSPBalance = stkPSPBalanceBN.toBigInt();
+
+    const totalPSPStaked =
+      (stkPSPBalance * bptBalanceOfStkPSPBpt * pspBalance) /
```
[the remark this answers is not in the document]
Comment (mwamedacen) [reply src/lib/staking/safety-module-helper.ts:75]: yep it is

[the remark this answers is not in the document]
Comment (mwamedacen) [reply src/lib/staking/safety-module-helper.ts:75]: actually to not suffer from too much precision loss, one have to minimise number of division. This is why this formula is expressed at it is.

The human readable version would be more like:

```javascript
stkPSPBalance/stkPSPBPtTotalSupply * bptBalanceOfStkPSPBpt/bptTotalSupply * pspBalance
```

**Pull request merged**